homeassistant/components/music_assistant/media_player.py

```diff
@@ -0,0 +1,751 @@
+    | MediaPlayerEntityFeature.PLAY
+    | MediaPlayerEntityFeature.PLAY_MEDIA
+    | MediaPlayerEntityFeature.VOLUME_STEP
+    | MediaPlayerEntityFeature.CLEAR_PLAYLIST
+    | MediaPlayerEntityFeature.BROWSE_MEDIA
+    | MediaPlayerEntityFeature.MEDIA_ENQUEUE
+    | MediaPlayerEntityFeature.MEDIA_ANNOUNCE
+    | MediaPlayerEntityFeature.SEEK
+)
+
+STATE_MAPPING = {
+    PlayerState.IDLE: STATE_IDLE,
+    PlayerState.PLAYING: STATE_PLAYING,
+    PlayerState.PAUSED: STATE_PAUSED,
+}
+
+QUEUE_OPTION_MAP = {
+    # map from HA enqueue options to MA enqueue options
+    # which are the same but just in case
+    MediaPlayerEnqueue.ADD: QueueOption.ADD,
+    MediaPlayerEnqueue.NEXT: QueueOption.NEXT,
+    MediaPlayerEnqueue.PLAY: QueueOption.PLAY,
+    MediaPlayerEnqueue.REPLACE: QueueOption.REPLACE,
+}
+
+ATTR_RADIO_MODE = "radio_mode"
+ATTR_MEDIA_ID = "media_id"
+ATTR_MEDIA_TYPE = "media_type"
+ATTR_ARTIST = "artist"
+ATTR_ALBUM = "album"
+ATTR_URL = "url"
+ATTR_USE_PRE_ANNOUNCE = "use_pre_announce"
+ATTR_ANNOUNCE_VOLUME = "announce_volume"
+ATTR_SOURCE_PLAYER = "source_player"
+ATTR_AUTO_PLAY = "auto_play"
+
+# pylint: disable=too-many-public-methods
+
+_MassPlayerT = TypeVar("_MassPlayerT", bound="MassPlayer")
+_R = TypeVar("_R")
+_P = ParamSpec("_P")
+
+
+def catch_musicassistant_error(
+    func: Callable[Concatenate[_MassPlayerT, _P], Awaitable[_R]],
+) -> Callable[Concatenate[_MassPlayerT, _P], Coroutine[Any, Any, _R | None]]:
+    """Check and log commands to players."""
+
+    @functools.wraps(func)
+    async def wrapper(
+        self: _MassPlayerT, *args: _P.args, **kwargs: _P.kwargs
+    ) -> _R | None:
+        """Catch Music Assistant errors and convert to Home Assistant error."""
+        try:
+            return await func(self, *args, **kwargs)
+        except MusicAssistantError as err:
+            raise HomeAssistantError(f"{err.__class__.__name__}: {err!s}") from err
+
+    return wrapper
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Music Assistant MediaPlayer(s) from Config Entry."""
+    mass = get_mass(hass, config_entry.entry_id)
+    added_ids = set()
+    if TYPE_CHECKING:
+        assert mass is not None
+
+    async def handle_player_added(event: MassEvent) -> None:
+        """Handle Mass Player Added event."""
+        if event.object_id in added_ids:
+            return
+        added_ids.add(event.object_id)
+        if TYPE_CHECKING:
+            assert event.object_id is not None
+        async_add_entities([MassPlayer(mass, event.object_id)])
+
+    # register listener for new players
+    config_entry.async_on_unload(
+        mass.subscribe(handle_player_added, EventType.PLAYER_ADDED)
+    )
+    # add all current players
+    for player in mass.players:
+        added_ids.add(player.player_id)
+        async_add_entities([MassPlayer(mass, player.player_id)])
```

Comment:
Only try to run `async_add_entities` once at setup, so collect all the entities in a list first